tensorflow/core/kernels/data/map_dataset_op_test.cc

```diff
@@ -0,0 +1,363 @@
+#include "tensorflow/core/kernels/data/iterator_ops.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/platform/test.h"
+#include "tensorflow/core/util/ptr_util.h"
+
+namespace tensorflow {
+namespace data {
+namespace {
+
+class MapDatasetOpTest : public DatasetOpsTestBase {
+ public:
+  static const char* kNodeName;
+  static const char* kOpName;
+
+  ~MapDatasetOpTest() override {
+    gtl::STLDeleteElements(&tensors_);
+    range_dataset_->Unref();
+  }
+
+ protected:
+  // Creates a new MapDataset dataset, meanwhile initializing the MapDataset op
+  // kernel and context internally.
+  DatasetBase* CreateMapDataset(DatasetBase* input_dataset,
+                                const string& func_name) {
+    map_inputs_.clear();
+    map_kernel_ =
+        CreateMapDatasetOpKernel<int64>(input_dataset->name(), func_name);
+
+    // Save the input dataset into a variant tensor as the input of MapDataset.
+    Tensor dataset_tensor(DT_VARIANT, TensorShape({}));
+    TF_CHECK_OK(StoreDatasetInVariantTensor(input_dataset, &dataset_tensor));
+    Variant variant = dataset_tensor.scalar<Variant>()();
+    TF_CHECK_OK(AddDatasetInputFromArray<Variant>(
+        &map_inputs_, map_kernel_->input_types(), TensorShape({}), {variant}));
+
+    TF_CHECK_OK(
+        CreateOpKernelContext(map_kernel_.get(), &map_inputs_, &map_context_));
+    TF_CHECK_OK(CheckOpKernelInput(*map_kernel_, map_inputs_));
+    DatasetBase* map_dataset;
+    TF_CHECK_OK(
+        CreateDataset(map_kernel_.get(), map_context_.get(), &map_dataset));
+    return map_dataset;
+  }
+
+ private:
+  // Creates a new MapDataset op kernel. The `input_dataset` parameter should be
+  // same with the node name of the input dataset for the method
+  // `CreateMapDataset()`.
+  template <typename T>
+  std::unique_ptr<OpKernel> CreateMapDatasetOpKernel(
+      const string& input_dataset, const string& func_name) {
+    DataTypeVector map_output_dtypes({tensorflow::DataTypeToEnum<T>::value});
+    gtl::ArraySlice<TensorShape> map_output_shapes({{}});
+    FunctionDefHelper::AttrValueWrapper func =
+        FunctionDefHelper::FunctionRef(func_name, {{"T", DT_INT64}});
+
+    map_node_def_ = test::function::NDef(kNodeName, kOpName, {input_dataset},
+                                         {{"f", func},
+                                          {"Targuments", {}},
+                                          {"output_shapes", map_output_shapes},
+                                          {"output_types", map_output_dtypes},
+                                          {"use_inter_op_parallelism", true},
+                                          {"preserve_cardinality", false}});
+    std::unique_ptr<OpKernel> map_op_kernel;
+    TF_CHECK_OK(CreateOpKernel(map_node_def_, &map_op_kernel));
+    return map_op_kernel;
+  }
+
+ protected:
+  DatasetBase* range_dataset_;
+  DatasetBase* map_dataset_;
+  std::unique_ptr<OpKernel> map_kernel_;
```

Comment:
I don't think that this needs to be a member variable. The unit tests should define these as local variables.